spring-cloud-dataflow-completion/src/main/java/org/springframework/cloud/dataflow/completion/CompletionUtils.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.dataflow.completion;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.springframework.cloud.dataflow.core.BindingProperties;
+import org.springframework.cloud.dataflow.core.ModuleCoordinates;
+import org.springframework.cloud.dataflow.core.ModuleDefinition;
+import org.springframework.cloud.dataflow.core.ModuleType;
+import org.springframework.cloud.dataflow.core.StreamDefinition;
+import org.springframework.cloud.stream.module.resolver.Coordinates;
+import org.springframework.util.Assert;
+
+/**
+ * Various utility methods used throughout the completion package.
+ *
+ * @author Eric Bottard
+ */
+public class CompletionUtils {
+
+	/**
+	 * The names of properties that may be added implicitly to all module definitions,
+	 * even though the user did not type them himself.
+	 */
+	static final Set<String> IMPLICIT_PARAMETER_NAMES = new HashSet<>();
+	static {
+		IMPLICIT_PARAMETER_NAMES.add(BindingProperties.INPUT_BINDING_KEY);
+		IMPLICIT_PARAMETER_NAMES.add(BindingProperties.OUTPUT_BINDING_KEY);
+	}
+
+	/**
+	 * Return the type(s) of a given module definition, in the context of code completion.
+	 */
+	static ModuleType[] inferType(ModuleDefinition moduleDefinition) {
```

Comment:
I had this discussion with @markfisher too.
Those methods are actually quite different. Why the one you mention operates in the context of a definition that we know (or at least assume) is 100% well formed (going from source-ish to sink-ish), this one -- although it receives an input that is well formed token-wise -- may receive an incomplete definition (we're here for code completion after all).

Hence, 

```
http | filter --expression=true | transform
```

is something we could get here, although `transform` is not a sink.

This method returns several values exactly for that reason: we're going to try to see if `transform` is a sink, or a processor